cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceQueryAdmin.java

```diff
@@ -83,16 +97,18 @@ public boolean apply(NamespaceMeta namespaceMeta) {
    */
   @Override
   public NamespaceMeta get(Id.Namespace namespaceId) throws Exception {
-    NamespaceMeta ns = nsStore.get(namespaceId);
-    if (ns == null) {
-      throw new NamespaceNotFoundException(namespaceId);
-    }
-    Principal principal = authenticationContext.getPrincipal();
-    Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
-    if (!Principal.SYSTEM.equals(principal) && !filter.apply(namespaceId.toEntityId())) {
-      throw new UnauthorizedException(principal, namespaceId.toEntityId());
+    try {
+      return namespaceMetaCache.get(namespaceId.toEntityId());
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof NamespaceNotFoundException) {
+        throw new NamespaceNotFoundException(namespaceId);
```

Comment:
Why create a new one? It'll lose the stacktrace. Should you just throw `e.getCause()` in this catch block?